### PR TITLE
Forest

### PR DESCRIPTION
**Pull request by FridljDa**
Added forest functionality. I will revisit smart default hyper parameters at a later stage. Special focus on backwards compatibility: any past code written using IHW should be unaffected.
points of deviation in original code:

- ihw-class for results has more slots and most getters behave differently depending on which stratification method was used
- multidimensional covariates are also allowed for quantile slicing (e.g. per-covariate bins). Hence covariate is allowed as matrix input and ihw.formula has been adjusted

### CODE REVIEW

File: R/helpers.R [hunk 1]
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
 	as.factor(ceiling( rfs* nbins))
 }
 
+#' Stratify hypotheses based on increasing value of the covariate 
+#'
+#'  Hypotheses are stratified into nbins different strata of (approximately) equal size based on
+#' increasing value of the covariate
+#' extension of `groups_by_filter` for multidimensional covariates, per-covariate quantiles
```
Comment (const-ae) [inline R/helpers.R]: You are using `` ` `` to indicate code. This relies on markdown formatting which Roxygen does not used by default. You can turn it on by adding `Roxygen: list(markdown = TRUE)` to the `DESCRIPTION` file (see for example https://github.com/const-ae/glmGamPoi/blob/master/DESCRIPTION). However as Nikos seems to be rather using the explicit notation (i.e. `\code{\link{rank}}`), so I would recommend that you adopt the same style.

File: R/helpers.R [hunk 2]
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
...
+#'
+#' @param covariates Numeric vector of ordinal covariates based on which the stratification will be done.
+#' @param nbins Integer, number of groups/strata into which p-values will be split based on covariate.
+#' @param ties.method Character specifying how ties are treated, see \code{\link{rank}} function.
+#' @param seed Integer, specifies random seed to be used when ties.method=="random".
```
Comment (const-ae) [inline R/helpers.R]: Wrap such code examples in `\code{ties.method=="random"}` so that they are rendered nicely.

File: R/helpers.R [hunk 3]
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
 	as.factor(ceiling( rfs* nbins))
 }
 
+#' Stratify hypotheses based on increasing value of the covariate 
+#'
+#'  Hypotheses are stratified into nbins different strata of (approximately) equal size based on
+#' increasing value of the covariate
+#' extension of `groups_by_filter` for multidimensional covariates, per-covariate quantiles
+#'
+#' @param covariates Numeric vector of ordinal covariates based on which the stratification will be done.
```
Comment (const-ae) [inline R/helpers.R]: Is `covariates` really a numeric vector? Line 58 (`covariates[, i]`) suggests that it should be two dimensional?

File: R/helpers.R [hunk 4]
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
...
+    old <- .Random.seed
+    on.exit({
+      .Random.seed <<- old
+    })
+    set.seed(as.integer(seed))
+  }
+  
+  nvar <- ncol(covariates)
+  groups <- lapply(seq_len(nvar), function(i) {
+    covariate_i <- covariates[, i]
```
Comment (const-ae) [inline R/helpers.R]: You have to be careful here. I assume you want your method to handle any kind of typical two-dimensional data structure (`matrix`, `data.frame`, and `tibble`). However tibbles behave a bit different: they don't automatically return a vector if you extract a single column:
``` r
tib <- tibble(a = letters[1:5], num = 1:5)
tib[,1]
#> # A tibble: 5 × 1
#>   a    
#>   <chr>
#> 1 a    
#> 2 b    
#> 3 c    
#> 4 d    
#> 5 e
tib[,1,drop=TRUE]
#> [1] "a" "b" "c" "d" "e"
```

File: R/helpers.R [hunk 5]
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
...
+  })
+  
+  groups <- do.call(cbind, groups)
+  groups <- as.data.frame(groups) 
+  
+  if(nvar == 1){
+    groups <- unname(unlist(groups))
+  }else{
+    groups <- apply(groups, 1, paste, collapse = "-") # base R equivalent of tidyr::unite
+  }
```
Comment (const-ae) [inline R/helpers.R:70]: Why did you convert `groups` into a data.frame if you anyway reduce the rows back to a vector?

File: R/helpers.R [hunk 6]
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
...
+  # convert to factor
+  groups <- as.factor(groups)
+  
+  return(groups)
```
Comment (const-ae) [inline R/helpers.R]: You can leave out the explicit `return`

File: R/ihw_class.R [hunk 7]
```diff
@@ -47,13 +51,17 @@ ihwResult <- setClass("ihwResult",
...
                    m_groups = "integer",
                    penalty = "character",
                    covariate_type = "character",
                    adjustment_type = "character",
                    reg_path_information = "data.frame",
-           		     solver_information= "list"))
+           		     solver_information= "list",
+           		     stratification_method = "character",
+           		     weight_matrices_forest = "list",
+           		     ntaus = "integer",
```
Comment (const-ae) [inline R/ihw_class.R]: I would avoid having greek letter variable names in public-facing methods or objects. They only make sense if you know which formula the letter refers to. I would thus recommend renaming all occurrences of `ntaus` to some more literal variable name like `n_thresholds` or `n_censor_thres`

File: R/ihw_class.R [hunk 8]
```diff
@@ -142,7 +162,12 @@ setMethod("weighted_pvalues", signature(object="ihwResult"),
 #---------------------------  covariate extraction ----------------------------------------------------------#
 
 covariates_ihwResult <- function(object){
-  object@df$covariate
+  covariate <- dplyr::select(object@df, dplyr::starts_with("covariate")) #TODO fix this
```
Comment (const-ae) [inline R/ihw_class.R]: `dplyr` is only listed as a suggested dependency. You cannot use it in regular code without checking that it is indeed available.

Comment (const-ae) [reply R/ihw_class.R]: You could replace the dplyr code by adapting the following snipped
``` r
dat <- data.frame(a_1 = 1:3, a_2 = 4:6, b_1 = letters[1:3], asdf = 0)
dat
#>   a_1 a_2 b_1 asdf
#> 1   1   4   a    0
#> 2   2   5   b    0
#> 3   3   6   c    0
sel_cols <- grepl("^a_", colnames(dat))
dat[,sel_cols]
#>   a_1 a_2
#> 1   1   4
#> 2   2   5
#> 3   3   6
```

File: R/ihw_class.R [hunk 9]
```diff
@@ -167,8 +192,12 @@ setMethod("covariate_type", signature(object="ihwResult"),
           covariate_type_ihwResult)
 
 #----------------- extract stratification variable------------------------------------------------------------#
-groups_factor_ihwResult <- function(object){
-	object@df$group
+groups_factor_ihwResult <- function(object){#
+  if(object@stratification_method == "forest"){
+    groups <- dplyr::select(object@df, dplyr::contains("tree"))
```
Comment (const-ae) [inline R/ihw_class.R]: See above!

File: R/plots.R [hunk 10]
```diff
@@ -27,90 +27,91 @@
...
-               ggplot2::geom_hex( bins = 100) +
-               ggplot2::geom_step( ggplot2::aes(x = minrk, y = -log10(pthresh), col = fold), data = df, direction = "hv")
+           ggplot2::ggplot(xdf, ggplot2::aes(x = rank(covariate), y = -log10(pvalue))) +
+             ggplot2::geom_hex( bins = 100) +
+             ggplot2::geom_step( ggplot2::aes(x = minrk, y = -log10(pthresh), col = fold), data = df, direction = "hv")
          },
          stop(sprintf("Invalid scale='%s'.", scale))
   ) 
-}  
+}
```
Comment (const-ae) [inline R/plots.R]: Please try to fix the white space in this file so that it gets easier to review

**Review by const-ae — COMMENTED** [review]
I went through parts of the code, and it looks quite solid to me. Of course, I cannot comment on the validity of the new random forest-based hypothesis weighting, but the code seems to be of high quality.

There are a few minor things that I have commented in-line; @nignatiadis feel free to correct me on anything that I have suggested :)


----------
Additional comments:

* What's up with the empty files?

File: R/ihw_class.R [hunk 11]
```diff
@@ -161,13 +161,10 @@ setMethod("weighted_pvalues", signature(object="ihwResult"),
...
 
-covariates_ihwResult <- function(object){
-  covariate <- dplyr::select(object@df, dplyr::starts_with("covariate")) #TODO fix this
-  if(ncol(covariate) == 1){
-    unlist(covariate)
-  }else{
-    as.matrix(covariate)
-  }
+covariates_ihwResult <- function(object){ 
+  covariate_names <- grepl("^covariate_", colnames(object@df))
```
Comment (const-ae) [inline R/ihw_class.R]: Sorry that was probably a bad example. The underscore `_` had no particular meaning in that regex!

File: R/helpers.R (hunk 5 above)
Comment (FridljDa) [reply R/helpers.R:70]: Thank you for pointing this out, it was obsolete at this point!

File: R/helpers.R (hunk 1 above)
Comment (FridljDa) [reply R/helpers.R]: Thank you, I changed it everywhere

File: R/helpers.R (hunk 2 above)
Comment (FridljDa) [reply R/helpers.R]: Thank you, I adressed this and also updated some documantation in `ihw_convex.R` accordingly

File: R/helpers.R (hunk 3 above)
Comment (FridljDa) [reply R/helpers.R]: No it is indeed not. I updated the documantation.

File: R/helpers.R (hunk 4 above)
Comment (FridljDa) [reply R/helpers.R]: I specified `drop=TRUE` in this line

File: R/helpers.R (hunk 6 above)
Comment (FridljDa) [reply R/helpers.R]: done

File: R/ihw_class.R (hunk 7 above)
Comment (FridljDa) [reply R/ihw_class.R]: I changed all `ntaus` to `n_censor_thres` in the whole project

File: R/plots.R (hunk 10 above)
Comment (FridljDa) [reply R/plots.R]: In the new commit, white space changes are finally not being displayed anymore (I was aware of this and had a few tries before)

**Pull request closed without merging**